### PR TITLE
Cloud control API with xml,attachments and file editing

### PR DESCRIPTION
**Pull request by doolse**
JS API for editing xml, attachments and files - the definition is in:
`cloudprovidersdk/controls/src/index.d.ts`

The typescript definitions for all the attachment types aren't complete but that will be more fleshed out when it comes to time to publish an sdk.

### CODE REVIEW

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/wizard/WizardItemEditor.scala [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+  override def hasBeenEdited(oldValue: Any, newValue: Any): Boolean = {
+    if (Objects.equals(oldValue, newValue)) false
+    else {
+      editDetected()
+      true
+    }
+  }
+
+  override def editDetected(): Unit = {
+    //    System.err.println("editDetected")
```
Comment (SammyIsConfused) [inline Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/wizard/WizardItemEditor.scala]: These `System.err.println` comments can probably be removed.

File: Source/Plugins/Core/com.equella.core/js/tsrc/cloudcontrol/index.ts [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+
+import {
+  ControlApi,
+  ItemCommand,
+  CloudControlRegister,
+  ControlParameters,
+  ItemState,
+  ItemCommandResponse,
+  FileEntries
+} from "../../../../../../../cloudprovidersdk/controls/src";
```
[the remark this answers is not in the document]
Comment (doolse) [reply Source/Plugins/Core/com.equella.core/js/tsrc/cloudcontrol/index.ts]: I've changed from a relative path to using a local node module (path is in package.json instead).

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/wizard/WizardApi.scala [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+import scala.collection.JavaConverters._
+import scala.util.Random
+
+object WizardApi {
+  lazy val editorMap = LegacyGuice.attachmentDeserializers.getBeanMap.asScala
+}
+
+@JsonTypeInfo(use = Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "command")
+@JsonSubTypes(
+  Array(
```
[the remark this answers is not in the document]
Comment (doolse) [reply Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/wizard/WizardApi.scala]: They aren't used anywhere in the server side, they're purely for JSON creation so don't need a constant.

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/wizard/WizardItemEditor.scala (hunk 1 above)
Comment (doolse) [reply Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/wizard/WizardItemEditor.scala]: Removed

**Pull request merged**